[Profiler] Fix Empty C Call Queue

sraikund16 opened the pull request:
Summary:
My commandeer of https://github.com/pytorch/pytorch/pull/150102

Based on description of PR it seems that we need to add C calls for each starting python event with a callable such that when the tracing exits we will have a matching enter for any given exit. It adds some unnecessary events at worst but prevents segfaults/failures. My PR just cleans up some refcount impl and logging.

Contributors: @arjun-choudhry 

Test Plan: Ran resnet test internally. Will check CI and ask reviewers to make sure it resolves their issues.

Differential Revision: D72207570

sraikund16 commented:
@pramodk @oraluben do you mind checking if this PR fixes the issue?

pramodk commented on torch/csrc/autograd/profiler_python.cpp:
```diff
@@ -901,6 +910,27 @@ void PythonTracer::recordCCall(
+  // To get the function object, you will need to look in the globals or the
+  // frame's f_globals
+  PyObject* func =
+      PyDict_GetItemString(frame->f_globals, name);
```
For Python 3.12, this is failing with:

```bash
pytorch/torch/csrc/autograd/profiler_python.cpp:927:33: error: invalid use of incomplete type ‘PyFrameObject’ {aka ‘struct _frame’}
  927 |       PyDict_GetItemString(frame->f_globals, name);
      |                                 ^~
In file included from /usr/include/python3.12/Python.h:42,
                 from pytorch/torch/csrc/autograd/profiler_python.cpp:13:
/usr/include/python3.12/pytypedefs.h:22:16: note: forward declaration of ‘PyFrameObject’ {aka ‘struct _frame’}
   22 | typedef struct _frame PyFrameObject;
```

IIUC, this needs to be changed for newer Python version ([>=3.11](https://docs.python.org/3.11/whatsnew/3.11.html)) compatibility.

sraikund16 committed:
[Profiler] Fix Empty C Call Queue (#150370)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/150370

My commandeer of https://github.com/pytorch/pytorch/pull/150102

Based on description of PR it seems that we need to add C calls for each starting python event with a callable such that when the tracing exits we will have a matching enter for any given exit. My diff just cleans up some refcount impl and logging.

Test Plan: Ran resnet test internally. Will check CI and ask reviewers to make sure it resolves their issues.

Differential Revision: D72207570

pramodk commented:
@sraikund16, thanks! I am able to build with latest change but I won't be able to test this at the moment / today. I am attaching here another simple test I saw failing with `nemo:25.02` container. In case you could test this with local build (it's standalone):

```python
import torch

def get_profiler():
    return torch.profiler.profile(
        activities=[torch.profiler.ProfilerActivity.CPU, torch.profiler.ProfilerActivity.CUDA],
        with_stack=True,
    )

def profile_tensor_ops():
    device = "cuda"
    with get_profiler() as prof:
        for _ in range(5):
            x = torch.randn(1000, 1000, device=device)
            y = torch.randn(1000, 1000, device=device)
            z = torch.matmul(x, y)
            torch.cuda.synchronize()
            prof.step()

    print(prof.key_averages().table(sort_by="self_cuda_time_total", row_limit=10))


if __name__ == "__main__":
    profile_tensor_ops()
```

and running as:

```bas
WORKSPACE_PATH=$(pwd)
docker run \
  --gpus all \
  -it \
  --rm \
  --ipc=host \
  --network=host \
  -v $WORKSPACE_PATH:$WORKSPACE_PATH \
  nvcr.io/nvidia/nemo:25.02 \
  python $WORKSPACE_PATH/test.py
```

was producing

```bash
   ....
  File "/usr/local/lib/python3.12/dist-packages/torch/profiler/profiler.py", line 777, in __exit__
    self.stop()
  File "/usr/local/lib/python3.12/dist-packages/torch/profiler/profiler.py", line 793, in stop
    self._transit_action(self.current_action, None)
  File "/usr/local/lib/python3.12/dist-packages/torch/profiler/profiler.py", line 836, in _transit_action
    action()
  File "/usr/local/lib/python3.12/dist-packages/torch/profiler/profiler.py", line 239, in stop_trace
    self.profiler.__exit__(None, None, None)
  File "/usr/local/lib/python3.12/dist-packages/torch/autograd/profiler.py", line 378, in __exit__
    self.kineto_results = _disable_profiler()
                          ^^^^^^^^^^^^^^^^^^^
RuntimeError: !stack.empty() INTERNAL ASSERT FAILED at "/opt/pytorch/pytorch/torch/csrc/autograd/profiler_python.cpp":982, please report a bug to PyTorch. Python replay stack is empty.
```

> Just one note: if you would test in local build, would be good to to verify if this test fails without this PR. Just for cross-checking as I was wondering if behavior changes a bit based on the different build configs. Nut I didn't get time to verify this thoroughly.

sraikund16 commented:
> @sraikund16, thanks! I am able to build with latest change but I won't be able to test this at the moment / today. I am attaching here another simple test I saw failing with `nemo:25.02` container. In case you could test this with local build (it's standalone):
> 
> ```python
> import torch
> 
> def get_profiler():
>     return torch.profiler.profile(
>         activities=[torch.profiler.ProfilerActivity.CPU, torch.profiler.ProfilerActivity.CUDA],
>         with_stack=True,
>     )
> 
> def profile_tensor_ops():
>     device = "cuda"
>     with get_profiler() as prof:
>         for _ in range(5):
>             x = torch.randn(1000, 1000, device=device)
>             y = torch.randn(1000, 1000, device=device)
>             z = torch.matmul(x, y)
>             torch.cuda.synchronize()
>             prof.step()
> 
>     print(prof.key_averages().table(sort_by="self_cuda_time_total", row_limit=10))
> 
> 
> if __name__ == "__main__":
>     profile_tensor_ops()
> ```
> 
> and running as:
> 
> ```vba
> WORKSPACE_PATH=$(pwd)
> docker run \
>   --gpus all \
>   -it \
>   --rm \
>   --ipc=host \
>   --network=host \
>   -v $WORKSPACE_PATH:$WORKSPACE_PATH \
>   nvcr.io/nvidia/nemo:25.02 \
>   python $WORKSPACE_PATH/test.py
> ```
> 
> was producing
> 
> ```shell
>    ....
>   File "/usr/local/lib/python3.12/dist-packages/torch/profiler/profiler.py", line 777, in __exit__
>     self.stop()
>   File "/usr/local/lib/python3.12/dist-packages/torch/profiler/profiler.py", line 793, in stop
>     self._transit_action(self.current_action, None)
>   File "/usr/local/lib/python3.12/dist-packages/torch/profiler/profiler.py", line 836, in _transit_action
>     action()
>   File "/usr/local/lib/python3.12/dist-packages/torch/profiler/profiler.py", line 239, in stop_trace
>     self.profiler.__exit__(None, None, None)
>   File "/usr/local/lib/python3.12/dist-packages/torch/autograd/profiler.py", line 378, in __exit__
>     self.kineto_results = _disable_profiler()
>                           ^^^^^^^^^^^^^^^^^^^
> RuntimeError: !stack.empty() INTERNAL ASSERT FAILED at "/opt/pytorch/pytorch/torch/csrc/autograd/profiler_python.cpp":982, please report a bug to PyTorch. Python replay stack is empty.
> ```
> 
> > Just one note: if you would test in local build, would be good to to verify if this test fails without this PR. Just for cross-checking as I was wondering if behavior changes a bit based on the different build configs. But I didn't get time to verify this thoroughly....
> 
> Edit 1: By the way, the reason for the `above note` is that I was further cross-checking above simple example with PyTorch containers:
> 
> * nvcr.io/nvidia/pytorch:25.01-py3, 2.6.0a0+ecf3bae40a.nv25.0
> * nvcr.io/nvidia/pytorch:25.02-py3, 2.7.0a0+ecf3bae40a.nv25.02
> * nvcr.io/nvidia/pytorch:25.02-py3, 2.7.0a0+7c8ec84dab.nv25.03
> 
> they all use Python 3.12.3 but only `25.01-py3` (`torch.__version__` -> `2.6.0a0+ecf3bae40a.nv25.01`) fails with the assert. And hence, initial conclusion that it's "only" Python version related is no longer true (?).
> 
> Edit 2: one common thing between failing `pytorch:25.01` and `nemo:25.02` is that `sys.version` for both is the same `'3.12.3 (main, Nov 6 2024, 18:32:19) [GCC 13.2.0]'` i.e. `GCC 13.2` whereas newer Pytorch are `3.12.3 (main, Feb 4 2025, 14:48:35) [GCC 13.3.0]` !

I don't have a machine with those virtual environments set up so I won't be able to test it myself. Since this PR is supposed to fix multiple issues, lets get it in and then we can do follow up on these other potential issues later.

sraikund16 commented:
@pytorchbot merge

pramodk commented:
> I don't have a machine with those virtual environments set up so I won't be able to test it myself. Since this PR is supposed to fix multiple issues, lets get it in and then we can do follow up on these other potential issues later.

👍  thank you @sraikund16! I will check this internally and update here if confirmed issue with GCC 3.12 vs 3.13 or something else.

arjun-choudhry commented:
@sraikund16 Can you please add me as a contributor in the PR?

clee2000 commented:
@pytorchbot revert -m "broke some profiler tests when building with debug asserts profiler/test_memory_profiler.py::TestMemoryProfiler::test_config_check [GH job link](https://github.com/pytorch/pytorch/actions/runs/14211763078/job/39822158330) [HUD commit link](https://hud.pytorch.org/pytorch/pytorch/commit/3ac5a499ddac701f607a9f7206f9bec8871e1cbb)" -c nosignal

sraikund16 commented:
> @sraikund16 Can you please add me as a contributor in the PR? Thanks

Added you to the PR description, I think you need to push to the branch itself to be considered a contributor on GH though. Let me know if there is another way

sraikund16 committed:
fix failing internal tests

sraikund16 commented:
@pytorchbot merge
